aten/src/ATen/native/TensorShape.cpp

```diff
@@ -49,7 +49,73 @@ Tensor & cat_out(Tensor & result, TensorList tensors, int64_t dim) {
   return at::_th_cat_out(result, tensors, dim);
 }
 
+static bool sizes_match_except(IntList s1, IntList s2, int64_t dim_except) {
+  if (s1.size() != s2.size()) {
+    return false;
+  }
+  for (int64_t i = 0; i < s1.size(); ++i) {
+    if (i != dim_except && s1[i] != s2[i]) {
+      return false;
+    }
+  }
+  return true;
+}
+
+static Tensor cat_sparse(TensorList tensors, int64_t dim) {
+  std::vector<Tensor> indices;
+  std::vector<Tensor> values;
+  int64_t sparse_dim = tensors[0].sparse_dim();
+  int64_t dense_dim = tensors[0].dense_dim();
+  AT_CHECK(dim < sparse_dim,
+           "Can't cat or stack tensors of sparse dim ", sparse_dim, "along non-sparse dimension ", dim);
+  IntList sizes = tensors[0].sizes();
+  for (size_t i = 0; i < tensors.size(); ++i) {
+    auto const &t = tensors[i];
+    AT_CHECK(t.is_sparse(),
+             "Can't cat dense tensor at position ", i, " with sparse tensor(s).");
+    AT_CHECK(t.sparse_dim() == sparse_dim && t.dense_dim() == dense_dim,
```

Comment:
I think this should be below the next check; it's much less obvious what is going on if you get this error message as opposed to the sizes one.